src/tools/wasm-ctor-eval.cpp

```diff
@@ -0,0 +1,445 @@
+        }
+        break;
+      }
+      case i64: {
+        switch (load->bytes) {
+          case 1: return load->signed_ ? Literal((int64_t)doLoad<int8_t>(addr)) : Literal((int64_t)doLoad<uint8_t>(addr));
+          case 2: return load->signed_ ? Literal((int64_t)doLoad<int16_t>(addr)) : Literal((int64_t)doLoad<uint16_t>(addr));
+          case 4: return load->signed_ ? Literal((int64_t)doLoad<int32_t>(addr)) : Literal((int64_t)doLoad<uint32_t>(addr));
+          case 8: return load->signed_ ? Literal((int64_t)doLoad<int64_t>(addr)) : Literal((int64_t)doLoad<uint64_t>(addr));
+          default: WASM_UNREACHABLE();
+        }
+        break;
+      }
+      case f32: return Literal(doLoad<float>(addr));
+      case f64: return Literal(doLoad<double>(addr));
+      default: WASM_UNREACHABLE();
+    }
+  }
+
+  void store(Store* store, Address addr, Literal value) override {
+    // store to the memory segments
+    switch (store->valueType) {
+      case i32: {
+        switch (store->bytes) {
+          case 1: doStore<int8_t>(addr, value.geti32()); break;
+          case 2: doStore<int16_t>(addr, value.geti32()); break;
+          case 4: doStore<int32_t>(addr, value.geti32()); break;
+          default: WASM_UNREACHABLE();
+        }
+        break;
+      }
+      case i64: {
+        switch (store->bytes) {
+          case 1: doStore<int8_t>(addr, (int8_t)value.geti64()); break;
+          case 2: doStore<int16_t>(addr, (int16_t)value.geti64()); break;
+          case 4: doStore<int32_t>(addr, (int32_t)value.geti64()); break;
+          case 8: doStore<int64_t>(addr, value.geti64()); break;
+          default: WASM_UNREACHABLE();
+        }
+        break;
+      }
+      // write floats carefully, ensuring all bits reach memory
+      case f32: doStore<int32_t>(addr, value.reinterpreti32()); break;
+      case f64: doStore<int64_t>(addr, value.reinterpreti64()); break;
+      default: WASM_UNREACHABLE();
+    }
+  }
+
+  void growMemory(Address /*oldSize*/, Address newSize) override {
+    throw FailToEvalException("grow memory");
+  }
+
+  void trap(const char* why) override {
+    throw FailToEvalException(std::string("trap: ") + why);
+  }
+
+private:
+  // TODO: handle unaligned too, see shell-interface
+
+  template <typename T>
+  T* getMemory(Address address) {
+    // if memory is on the stack, use the stack
+    if (address >= instance->STACK_START) {
+      Address relative = address - instance->STACK_START;
+      if (relative + sizeof(T) > instance->STACK_SIZE) {
+        throw FailToEvalException("stack usage too high");
+      }
+      // in range, all is good, use the stack
+      return (T*)(&instance->stack[relative]);
+    }
+
+    // otherwise, this must be in the singleton segment. resize as needed
+    if (wasm->memory.segments.size() == 0) {
+      std::vector<char> temp;
+      wasm->memory.segments.push_back(
+        Memory::Segment(
+          wasm->allocator.alloc<Const>()->set(Literal(int32_t(0))),
```

Comment:
That's a good point, I hadn't considered it from a declarative perspective. Looking at it again, I think it's kind of a wash. But, here are some thoughts:

1. The biggest thing that jumped out is the very long expression that allocates the `Const` with a literal value of 0. A helper function like `Const* newInt32Const(int32_t value);` would work wonders here.
2. We have a `temp` variable as part of the expression, which causes the reader's eyes to have to backtrack to the definition above anyway. We can replace that with an anonymous rvalue, just an inline `std::vector<char>()`, which is in keeping with the ast-ey way.
3. We're mixing object construction and pushing that object to a vector. Which is clearly correct if we're doing something like `vec.push_back(4)`, but becomes a stylistic thing for larger expressions.
4. C++ isn't the greatest at expressing DSLs like that. In something like Python you could do:
```python
segments.append(
  Memory.Segment(
    offset=Const(Literal(0)),
    init=temp
  )
)
```
and if you had a bigger more complex expression it'd be easier to follow:
```python
foo = Big(
  nasty=Nested(
    expression=That(),
    we=Clearly(
      have=All(the),
      right=Vars()
    ),
    for=Because()
  ),
  the=Language(supports=it)
)
```
, vs in C++
```c++
foo = Well(
  this->Definitely(
    Compiles(),
    but(
      From(a),
      Readers,
    ),
    perspective,
    It(can),
  Be(Hard(
    to(Know()),
    what
  ),
  all(this),
  means
)
```
, so something in the style of
```c++
auto segment = Memory::Segment(offset, init);
```
is more typical.

The main goal I have is "how can we minimize the mental stack needed for a reader to parse this expression," so by apply the technique of intermediate variables we divide the expression into smaller chunks that a reader can process individually.
There's more than one way to do that though, and something like
```c++
wasm->memory.segments.push_back(
  Memory::Segment(newInt32Const(0), std::vector<char>())
);
```
is quite readable, and has an AST-ey flavor to it.